src/main/java/com/google/devtools/build/lib/bazel/bzlmod/IndexRegistry.java

```diff
@@ -182,17 +187,36 @@ private Optional<byte[]> doGrabFile(
           "Cannot fetch a file without a checksum in ENFORCE mode. This is a bug in Bazel, please "
               + "report at https://github.com/bazelbuild/bazel/issues/new/choose.");
     }
+
+    URL url = new URL(rawURL);
+    // Don't try to read the registry URL from the vendor directory in the following cases:
+    // 1. The vendor directory is not set, which means vendor mode is disabled.
+    // 2. The checksum is not present, which means the URL is not vendored or the vendored content is out-dated.
+    // 3. The URL starts with "file:", which means it's a local file and isn't vendored.
+    // Otherwise, check if the URL is vendored and read the registry file from the vendor directory.
+    if (vendorDir.isPresent() && checksum.isPresent() && !url.getProtocol().equals("file")) {
+      VendorUtil vendorUtil = new VendorUtil(vendorDir.get());
```

Comment:
Could we make this a field instead of `vendorDir`?

Reply:
Totally, thanks for the suggestion!